### PR TITLE
Fix #14239: Correctly synchronize parallel image downloads to avoid possible corruption cause

### LINKED ISSUE #14239: Parallel pull from V2 registry may lead to corruption of shared ancestor images
### Description

Pulling images with common ancestors from a V2 private registry seems to have some probability of corrupting the images, and the issue seems to stem from the client side. Serializing download of all images made the problem go away, but is not a good user solution because (A) it cannot be done correctly outside of the docker daemon and (B) it slows downloads which could otherwise be done in parallel.

Due to the severity of this issue, we have been unable to use the docker registry in production, for fear of corrupted images leading to hard-to-trace failures.

### Bug report

*List the steps to reproduce the issue:*
1. Execute many "docker pull" requests in parallel against images in a V2 private registry which share several common ancestors.
2. docker run all of them after downloads have been completed.

*Describe the results you received:*
Nondeterministically, this produces corrupt images. If ubuntu is corrupted, for instance, we may receive (during build time of subsequent images):
`System error: exec: "/bin/bash": stat /bin/bash: no such file or directory`

In other cases, we've seen aufs mount failed exceptions while running images which had been fully downloaded and built without error, but unfortunately I do not have one of these errors recorded.

*Describe the results you expected:*
Pulling and building should deterministically work with parallel operations.

*Provide additional info you think is important:*
We worked around the issue on our test machines and development environment by serializing all "docker pull" operations behind a global lock. This caused both the /bin/bash error and the aufs mount failed error to go away, and the build returned to relative stability.

*docker version:*
1.6.2

*docker info:*
Containers: 88
Images: 1715
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 1891
 Dirperm1 Supported: false
Execution Driver: native-0.2
Kernel Version: 3.11.0-26-generic
Operating System: Ubuntu 13.10
CPUs: 4
Total Memory: 15.38 GiB
WARNING: No swap limit support

*uname -a:*
Linux kitties 3.11.0-26-generic #45-Ubuntu SMP Tue Jul 15 04:02:06 UTC 2014 x86_64 x86_64 x86_64 GNU/Linux

*Provide additional environment details (AWS, VirtualBox, physical, etc.):*
Reproducible on AWS, boot2docker on Mac, and Linux laptop.

### PR DESCRIPTION
**Pull request by aarondav**
Pulling images with common ancestors from a V2 private registry seems to have some probability of corrupting the images, and the issue seems to stem from the client side. Serializing download of all images made the problem go away, but is not a good user solution because (A) it cannot be done correctly outside of the docker daemon and (B) it slows downloads which could otherwise be done in parallel.

This patch revises the parallelism story of the V2 docker pull to avoid releasing the pool lock from images before the download AND extraction has been completed. This means that the V2 pull has the same behavior as the old V1 pull (which performed both download and extraction in the same method, holding the image lock).

I would like to add an integration test, but the reproduction case is very expensive to perform in a test that is run frequently (download several images which share multiple common ancestors in parallel). Manual testing on our workload with this patch has not been able to reproduce the issue after several attempts, whereas previously it would happen with very high frequency.

### CODE REVIEW

File: graph/pull.go [hunk 1]
```diff
@@ -635,24 +638,38 @@ func (s *TagStore) pullV2Tag(r *registry.Session, out io.Writer, endpoint *regis
...
 			if err := <-d.err; err != nil {
-				return false, err
+				logrus.Debugf("Failed to download %s with err %s", d.img.ID, err)
+				lastErr = err
 			}
 		}
-		if d.downloaded {
-			// if tmpFile is empty assume download and extracted elsewhere
+
+		if d.tmpFile != nil {
```
Comment (aarondav) [inline graph/pull.go]: This seemed to be a bug before; if all downloads were not successful, we would leak the temp downloaded files.

File: graph/pull.go [hunk 2]
```diff
@@ -635,24 +638,38 @@ func (s *TagStore) pullV2Tag(r *registry.Session, out io.Writer, endpoint *regis
...
 
+	var lastErr error
 	var tagUpdated bool
 	for i := len(downloads) - 1; i >= 0; i-- {
 		d := &downloads[i]
 		if d.err != nil {
 			if err := <-d.err; err != nil {
-				return false, err
+				logrus.Debugf("Failed to download %s with err %s", d.img.ID, err)
+				lastErr = err
```
Comment (aarondav) [inline graph/pull.go]: Behavior of this loop has been changed; now, we do not exit early if one of the downloads failed, but rather wait for all parallel downloads to fail. This allows us to correctly clean up the temp files and unlock the images.

**Comment by estesp** [comment]
I can't tell if this is a different root problem than what is being fixed by #14193 but @stevvooe can probably take a look and make an assessment.

**Comment by aarondav** [comment]
This patch does corroborate -- image corruption as a symptom should only have been caused by Graph.Register() not being threadsafe, so I'm glad to see #14193. However, there are other, subtler and less critical issues that could arise without the enhanced concurrency protection in this patch; namely, spurious failures on threads that were woken up after the image has been downloaded but is not yet registered.

File: graph/pull.go [hunk 3]
```diff
@@ -545,7 +545,11 @@ func (s *TagStore) pullV2Tag(r *registry.Session, out io.Writer, endpoint *regis
 		tmpFile    *os.File
 		length     int64
 		downloaded bool
-		err        chan error
+		// set to true if another thread is attempting to download the same image.
+		// if true, waitChannel will be closed when the download is complete.
+		pullAlreadyInProgress bool
```
Comment (stevvooe) [inline graph/pull.go]: This needs to be protected by a mutex if it is to do what is suggested.

File: graph/pull.go [hunk 4]
```diff
@@ -579,16 +583,15 @@ func (s *TagStore) pullV2Tag(r *registry.Session, out io.Writer, endpoint *regis
 		downloadFunc := func(di *downloadInfo) error {
 			logrus.Debugf("pulling blob %q to V1 img %s", sumStr, img.ID)
 
-			if c, err := s.poolAdd("pull", "img:"+img.ID); err != nil {
-				if c != nil {
+			if di.pullAlreadyInProgress {
+				if di.waitChannel != nil {
```
Comment (stevvooe) [inline graph/pull.go]: This test is not threadsafe.

**Comment by stevvooe** [comment]
@aarondav As it stands, this PR is pretty "racy". It may better to break down a few of the fixes into separate PRs, so we can get the code right. #14193 is not a complete solution, but I'd like to at least get the resource leak fixes in rather than holding them up on this PR.

**Comment by aarondav** [comment]
Perhaps I misunderstand the go concurrency model, but the "downloadInfo" object is scoped to inside a single pullTagV2 call, and so it is not being concurrently modified. It is being passed from the pullTagV2 method into the goroutine in a one-way, serialized communication.

**Comment by stevvooe** [comment]
@aarondav Would you mind handling the resource fix in another PR? This PR will take more time to understand.

**Comment by aarondav** [comment]
Of course. I will try to do the separation soon. Thanks for the quick
feedback!

**Comment by aarondav** [comment]
I have created #14389 which simply does the cleanup. There I also changed the register part of the loop into a function, which should significantly simplify the changes here (rather than having several if statements, one defer will do).

**Comment by aarondav** [comment]
Second commit just has diff from other PR:
https://github.com/aarondav/docker/commit/e6d45a42d856073ecbc56cd15367ec6f77a5d5e9
which is hopefully now a significantly easier-to-verify change.

**Comment by cpuguy83** [comment]
Can we close this for the time being since it seems to require ##14389 first?

**Pull request closed without merging**